### PR TITLE
Enhance error when input files are not found

### PR DESCRIPTION
**Pull request by schlunma**
This addresses an issue of @ledm's presentation at the workshop: Right now, the tool simply outputs

```
No input files found for variable {...}
```

if it does not find input data. To get an idea of what went wrong, user's have to look at the debug log. Also, I think the attribute `filename` in the variable dictionary is very confusing, since it does only indicate the location of the file after preprocessing, which is definitely not useful here.

The main changes are in the following code snippet. The text of the actual error messages themselves is of course open for discussion.

https://github.com/ESMValGroup/ESMValCore/blob/b9d82b05913709d4b322c3684b75818cd0d2e152/esmvalcore/_recipe_checks.py#L95-L110

### CODE REVIEW

**Comment by ledm** [comment]
Good work @schlunma. The PR runs as expected! I ran it and it outputs better error messages.

I was thinking that the message:
```
Looked for files matching ['msftmyz_Omon_CanESM25_historical_r1i1p1_*.nc'] in ['/users/modellers/ledm/workspace/ESMValToolTest/KIT_data/']
```
would be better if the filename and directory were combined. Ie:
```
Looked for files matching: '/users/modellers/ledm/workspace/ESMValToolTest/KIT_data/msftmyz_Omon_CanESM25_historical_r1i1p1_*.nc'
```

You could then copy and paste it into the command line and figure out whats the problem. Does that make sense?

**Comment by schlunma** [comment]
> I was thinking that the message:
> 
> ```
> Looked for files matching ['msftmyz_Omon_CanESM25_historical_r1i1p1_*.nc'] in ['/users/modellers/ledm/workspace/ESMValToolTest/KIT_data/']
> ```
> 
> would be better if the filename and directory were combined. Ie:
> 
> ```
> Looked for files matching: '/users/modellers/ledm/workspace/ESMValToolTest/KIT_data/msftmyz_Omon_CanESM25_historical_r1i1p1_*.nc'
> ```
> 
> You could then copy and paste it into the command line and figure out whats the problem. Does that make sense?

Yeah, makes perfectly sense, this is way better :grin: I will change it right now.

File: esmvalcore/_recipe_checks.py [hunk 1]
```diff
@@ -90,10 +91,25 @@ def variable(var, required_keys):
...
+def data_availability(input_files, var, dirnames, filenames):
     """Check if the required input data is available."""
     if not input_files:
-        raise RecipeError("No input files found for variable {}".format(var))
+        var.pop('filename', None)
+        logger.error("No input files found for variable %s", var)
+        if dirnames and filenames:
+            patterns = itertools.product(dirnames, filenames)
+            patterns = [os.path.join(d, f) for (d, f) in patterns]
+            logger.error("Looked for files matching\n%s", "\n".join(patterns))
```
Comment (ledm) [inline esmvalcore/_recipe_checks.py]:
```suggestion
            logger.error("Looked for files matching: %s", "\n".join(patterns))
```

File: esmvalcore/_recipe_checks.py [hunk 2]
```diff
@@ -90,10 +91,25 @@ def variable(var, required_keys):
...
+        elif dirnames and not filenames:
+            logger.error(
+                "Looked for files in %s, but did not find any file pattern "
+                "to match against", dirnames)
+        elif filenames and not dirnames:
+            logger.error(
+                "Looked for files matching %s, but did not find any existing "
+                "input directory", filenames)
+        logger.error("Set 'log_level' to 'debug' to get more information")
```
Comment (ledm) [inline esmvalcore/_recipe_checks.py]: Is this comment necessary?

Comment (schlunma) [reply esmvalcore/_recipe_checks.py]: Yes, I think so. For example. if `dirnames` is an empty list (which happens when the directory does not exist), the user gets no information about the directory ESMValTool was trying to find. Only the debug log shows this.

File: esmvalcore/_recipe_checks.py (hunk 1 above)
Comment (schlunma) [reply esmvalcore/_recipe_checks.py]: I agree that this is nicer if `patterns` only contains one element, but if it contains more it's not really pretty. I change it to respect both cases.

**Comment by mattiarighi** [comment]
Can we merge this?

**Comment by bouweandela** [comment]
No, I want to have a look at the `intake` library (see #31) before doing major work on data_finder.py.

**Comment by valeriupredoi** [comment]
@bouweandela and meself had a discussion at the Workshop about `intake` and we both agreed that it'd be a good thing but not just yet since it'll involve a major overhaul of all the data finding and logging functionalities; so I say let's get this in since it'll be very helpful especially in conunction with the dryrun capability #307

**Comment by valeriupredoi** [comment]
quite a few conflicts tho, wouldn't want to be the one fixing them :rofl:

**Comment by valeriupredoi** [comment]
@schlunma would you be a peach and fix the conflicts pls, then I'll approve and it'll be good to go :beer:

**Comment by schlunma** [comment]
Ahhh, now you were too fast, I wanted to add a test for the `Check.data_availability` feature :grimacing: I also haven't tested it yet, I'm really not sure if that broke anything :smile:

**Comment by valeriupredoi** [comment]
should be fine - can you add the test in a separate PR? That makes more sense since this dealt with printing messages only :beer:

**Pull request merged**